libs/langchain/langchain/retrievers/parent_document_retriever.py

```diff
@@ -0,0 +1,139 @@
+        # It should create documents smaller than the parent
+        child_splitter = RecursiveCharacterTextSplitter(chunk_size=400)
+        # The vectorstore to use to index the child chunks
+        vectorstore = Chroma(embedding_function=OpenAIEmbeddings())
+        # The storage layer for the parent documents
+        store = InMemoryStore()
+
+        # Initialize the retriever
+        retriever = ParentDocumentRetriever(
+            vectorstore=vectorstore,
+            docstore=store,
+            child_splitter=child_splitter,
+            parent_splitter=parent_splitter,
+        )
+    """
+
+    vectorstore: VectorStore
+    """The underlying vectorstore to use to store small chunks
+    and their embedding vectors"""
+    docstore: BaseStore[str, Document]
+    """The storage layer for the parent documents"""
+    child_splitter: TextSplitter
+    """The text splitter to use to create child documents."""
+    id_key: str = "doc_id"
+    """The key to use to track the parent id. This will be stored in the
+    metadata of child documents."""
+    parent_splitter: Optional[TextSplitter] = None
+    """The text splitter to use to create parent documents.
+    If none, then the parent documents will be the raw documents passed in."""
+
+    def get_relevant_documents(
+        self,
+        query: str,
+        *,
+        callbacks: Callbacks = None,
+        tags: Optional[List[str]] = None,
+        metadata: Optional[Dict[str, Any]] = None,
+        **kwargs: Any,
+    ) -> List[Document]:
+        sub_docs = self.vectorstore.similarity_search(query)
+        # We do this to maintain the order of the ids that are returned
+        ids = []
+        for d in sub_docs:
+            if d.metadata[self.id_key] not in ids:
+                ids.append(d.metadata[self.id_key])
+        docs = self.docstore.mget(ids)
+        return [d for d in docs if d is not None]
+
+    def add_documents(
+        self,
+        documents: List[Document],
+        ids: Optional[List[str]],
+        add_to_docstore: bool = True,
+    ) -> None:
+        """Adds documents to the docstore and vectorstores.
+
+        Args:
+            documents: List of documents to add
+            ids: Optional list of ids for documents. If provided should be the same
+                length as the list of documents. Can provided if parent documents
+                are already in the document store and you don't want to re-add
+                to the docstore. If not provided, random UUIDs will be used as
+                ids.
+            add_to_docstore: Boolean of whether to add documents to docstore.
+                This can be false if and only if `ids` are provided. You may want
+                to set this to False if the documents are already in the docstore
+                and you don't want to re-add them.
+        """
+        if self.parent_splitter is not None:
+            documents = self.parent_splitter.split_documents(documents)
```

Comment:
The parent splitter causes inconsistent semantics with `ids` -- this part should probably not be an option provided by the retriever and instead to a utility that the user has to invoke before using the splitter

Reply:
Agreed